### ISSUE TITLE
'forward_normalize' function 

### ISSUE THREAD

**Issue by talrub** [Ping-C/optimizer#5]
Hi,
At section 4.2 ("Results on 2-Class CIFAR-10/MNIST") in the paper you say: "we compare the performance of the models across different train loss levels **after the model's weights have been normalized**".
'LeNetModels' class contains 2 functions that are related to normalization:  'normalize' and 'forward_normalize'.
You don't use 'normalize' function during the training with G&C optimizer , but you do use 'forward_normalize'.
I have some questions regarding this function:

1. Can you explain what's the idea behind normalizing the forward pass by the 'cum_norm' of the model's weights? I don't see how this function is equivalent to normalizing the model's weights.

2. I saw that you are not using 'Softmax' at the end of your model as usually done in classification problem.  Why is that? Is the 'forward_normalize' function is a substitute for the 'Softmax'? I am guessing that applying 'Softmax' at the end of the forward pass and using 'forward_normalize' will result in outputs that are close to zero.

3. In line 509 at 'train_distributed.py' you pass 'forward_normalize' as the forward pass function for the train loss and accuracy calculation, but in line 687 you pass 'forward' as the forward pass function for the test train loss and accuracy calculation. Can you explain this difference?

Thanks!

**Comment by Ping-C** [comment]
Hey Tal,

If we focus on a single layer $ReLU(Wx)$, after normalization of the weights, we get $ReLU(\frac{W}{||W||_2}x)$, which is equivalent to $\frac{1}{||W||_2}ReLU(Wx)$. The logic applies to all layers, so we can simply divide the final output by the cumulative norm.

When calculating loss with cross entropy, the loss function already implicitly do a softmax of the final output (see https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html). 

`train_distributed.py` doesn't have line 687, but if you are referring to line 577. The line is there to test that the train accuracy is 100%, which normalization doesn't any impact. If you are referring to another line, I can look into it.

Hope this helps!

**Comment by talrub** [comment]
Hi,

Regarding your explanation for the normalization of the weights - i can see how it makes sense when using 'ReLU' activation but what should i do if i want to use 'sigmoid' function in my model ( which is common in RNN model for example)? In that case should i manually normalize the model weights before the forward pass ( as in 'normalize' function but without normalization of biases)?

I am sorry for my confusion regarding train_distributed.py code lines, i was referring to line 578 where you calculate test loss and accuracy by passing 'new_models' to 'calculate_loss_acc' and not 'new_models.forward_normalize' as i would expect after looking at line 509.

Thanks!

**Comment by Ping-C** [comment]
For sigmoid function, you cannot use the division by cumulative norm because sigmoid does not have the stated property. In your case I would simply do weight normalization.

For 578, I am not doing forward normalize because I was only using train accuracy, where normalization doesn’t have any impact.
